train.c

```diff
@@ -0,0 +1,464 @@
+    }
+    // exploit
+    int max_act = -1;
+    float max_q = -DBL_MAX;
+    printf("[ ");
+    for (int i = 0; i < move_cnt; i++) {
+        char *try_table = malloc(sizeof(char) * N_GRIDS);
+        memcpy(try_table, table, sizeof(char) * N_GRIDS);
+        try_table[available_moves[i]] = 'O';
+        printf("%f ", state_value_player_O[hash(try_table)]);
+        if (state_value_player_O[hash(try_table)] > max_q) {
+            max_q = state_value_player_O[hash(try_table)];
+            max_act = available_moves[i];
+        }
+    }
+    printf(" ]\n");
+    epsilon *= GAMMA;
+    printf("exploit %d\n", max_act);
+    return max_act;
+}
+
+int get_action_X(char *table)
+{
+    int move_cnt = 0;
+    int *available_moves = get_available_moves(table, &move_cnt);
+    if (RAND_UNIFORM < epsilon) {
+        printf("expore %d\n", available_moves[rand() % move_cnt]);
+        return available_moves[rand() % move_cnt];
+    }
+    int max_act = -1;
+    float max_q = -DBL_MAX;
+    printf("[ ");
+    for (int i = 0; i < move_cnt; i++) {
+        char *try_table = malloc(sizeof(char) * N_GRIDS);
+        memcpy(try_table, table, sizeof(char) * N_GRIDS);
+        try_table[available_moves[i]] = 'X';
+        printf("%f ", state_value_player_X[hash(try_table)]);
+        if (state_value_player_X[hash(try_table)] > max_q) {
+            max_q = state_value_player_X[hash(try_table)];
+            max_act = available_moves[i];
+        }
+    }
+    printf(" ]\n");
+    epsilon *= GAMMA;
+    printf("exploit %d\n", max_act);
+    return max_act;
+}
+
+void update_state_value_O(int cur_state_hash, int nxt_state_hash)
+{
+    int nxt_reward = get_score(hash_to_table(nxt_state_hash), 'O');
+    state_value_player_O[cur_state_hash] =
+        (1 - LEARNING_RATE) * state_value_player_O[cur_state_hash] +
+        LEARNING_RATE *
+            (nxt_reward + GAMMA * state_value_player_O[nxt_state_hash]);
+}
+
+void update_state_value_X(int cur_state_hash, int nxt_state_hash)
+{
+    int nxt_reward = get_score(hash_to_table(nxt_state_hash), 'X');
+    state_value_player_X[cur_state_hash] =
+        (1 - LEARNING_RATE) * state_value_player_X[cur_state_hash] +
+        LEARNING_RATE *
+            (nxt_reward + GAMMA * state_value_player_X[nxt_state_hash]);
+}
+
+void train(int iter)
+{
+    char *table = malloc(sizeof(char) * N_GRIDS);
+    if (!table)
+        exit(1);
+    memset(table, ' ', sizeof(char) * N_GRIDS);
+    char turn = 'X';
+    if (iter % 2 == 0)
+        turn = 'O';
+    char playerO = 'O';
+    char playerX = 'X';
+
+    while (1) {
+        char win = check_win(table);
+        if (win == 'D') {
+            draw_board(table);
+            printf("It is a draw!\n");
+            break;
+        } else if (win != ' ') {
+            draw_board(table);
+            printf("%c won!\n", win);
+            break;
+        }
+        int cur_state_hash = hash(table);
+        if (turn == playerO) {
+            int move = get_action_O(table);
```

Comment:
It seems that you use a random player as the AI's opponent when training, It does not make sense. you can let the AI player play against itself or the original AI instead.